shadowsocks/salt_generator.go

```diff
@@ -44,61 +44,58 @@ var RandomSaltGenerator SaltGenerator = randomSaltGenerator{}
+// we generate an unmarked random salt.
+const minEntropy = 16
+
 // Constant to identify this marking scheme.
 var serverSaltLabel = []byte("outline-server-salt")
 
 // NewServerSaltGenerator returns a SaltGenerator whose output is apparently
 // random, but is secretly marked as being issued by the server.
 // This is useful to prevent the server from accepting its own output in a
 // reflection attack.
-func NewServerSaltGenerator(cipher shadowaead.Cipher) (ServerSaltGenerator, error) {
-	saltSize := cipher.SaltSize()
-	zeroSalt := make([]byte, saltSize)
-	encrypter, err := cipher.Encrypter(zeroSalt)
-	if err != nil {
-		return ServerSaltGenerator{}, err
-	}
-	return ServerSaltGenerator{saltSize, encrypter}, nil
+func NewServerSaltGenerator(secret string) *ServerSaltGenerator {
+	// Shadowsocks already uses HKDF-SHA1 to derive the AEAD key, so we use
+	// the same derivation with a different "info" to generate our HMAC key.
+	keySource := hkdf.New(crypto.SHA1.New, []byte(secret), nil, serverSaltLabel)
+	// The key can be any size, but matching the block size is most efficient.
+	key := make([]byte, crypto.SHA1.Size())
+	io.ReadFull(keySource, key)
+	return &ServerSaltGenerator{key}
 }
 
-func (sg ServerSaltGenerator) splitSalt(salt []byte) (prefix, mark []byte) {
-	prefixLen := sg.saltSize - markLen
-	prefix = salt[:prefixLen]
-	mark = salt[prefixLen:]
-	return
+func (sg *ServerSaltGenerator) splitSalt(salt []byte) (prefix, mark []byte) {
+	prefixLen := len(salt) - markLen
+	return salt[:prefixLen], salt[prefixLen:]
 }
 
 // getTag takes in a salt prefix and returns the tag.
-// len(prefix) must be saltSize - markLen, which must be at least nonceSize.
-// prefix must be random to avoid nonce reuse.
-func (sg ServerSaltGenerator) getTag(prefix []byte) []byte {
-	// Only the first nonceSize bytes are used to compute the tag. In the event
-	// of a nonce collision (p=2^-33 after 2^32 messages for nonceSize==12),
-	// the only effect will be to reveal a pattern in the handshakes, not
-	// to reuse the same nonce on different inputs (which can cause more
-	// serious problems: https://www.imperialviolet.org/2015/05/16/aeads.html).
-	nonce := prefix[:sg.encrypter.NonceSize()]
-	return sg.encrypter.Seal(nil, nonce, nil, serverSaltLabel)
+func (sg *ServerSaltGenerator) getTag(prefix []byte) []byte {
+	// Use HMAC-SHA1, even though SHA1 is broken, because HMAC-SHA1 is still
+	// secure, and we're already using HKDF-SHA1.
+	hmac := hmac.New(crypto.SHA1.New, sg.key)
+	hmac.Write(prefix) // Hash.Write never returns an error.
+	return hmac.Sum(nil)
 }
 
 // GetSalt returns an apparently random salt that can be identified
 // as server-originated by anyone who knows the Shadowsocks key.
-func (sg ServerSaltGenerator) GetSalt(salt []byte) error {
-	if len(salt) != sg.saltSize {
-		return fmt.Errorf("Wrong salt size: %d != %d", len(salt), sg.saltSize)
+func (sg *ServerSaltGenerator) GetSalt(salt []byte) error {
+	if len(salt)-markLen < minEntropy {
```

Comment:
OK.  I moved it into the constructor, to avoid having to expose `markLen` and `minEntropy` to cipher_entry.go.  This also triggered some other changes to how salt sizes are checked, and to the tests.